Swift Implementation for underlying Queue::writeBuffer

nmahendru opened the pull request:
#### 26880c142f2708a75e519e311edf1b99e9ab208b
<pre>
Swift Implementation for underlying Queue::writeBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=280715">https://bugs.webkit.org/show_bug.cgi?id=280715</a>
<a href="https://rdar.apple.com/136672126">rdar://136672126</a>

Reviewed by Mike Wyrzykowski.

Re-implement Queue::writeBuffer using a Swift implementation.
It is still using some underlying utilities from the C++ implementation
but that can only be moved over once Swift implements the whole class and has ownership
of underlying objects.

Testing:
Manual at desk builds with --webGpuSwift and running the Buffer and Queue tests.
Then Run writeBuffer.html test.

* Source/WTF/wtf/RefCounted.h:
(retainRefCounted): Deleted.
(releaseRefCounted): Deleted.
* Source/WTF/wtf/RetainReleaseSwift.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(retainThreadSafeRefCounted): Deleted.
(releaseThreadSafeRefCounted): Deleted.
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::retainThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
(WTF::releaseThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
(retainBuffer):
(releaseBuffer):
* Source/WebGPU/WebGPU/Device.h:
(retainDevice):
(releaseDevice):
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::blitCommandEncoder):
(retainQueue):
(releaseQueue):
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):
(writeBuffer(_:buffer:offset:data:)): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: <a href="https://commits.webkit.org/284638@main">https://commits.webkit.org/284638@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04059647ae7544b19b8a1027779dee1835feb0a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49271 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73955 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21028 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20879 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55476 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13947 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72936 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35957 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41562 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17704 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19405 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62986 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63493 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18048 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75670 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69116 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17282 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63160 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60345 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63091 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15542 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11109 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4717 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90898 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45074 "Built successfully") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16148 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46148 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47419 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -155,12 +155,13 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
 
 } // namespace WebGPU
 
+// FIXME: rdar://137074301
```
Not really too important, but should we leave a WebKit bug link too and not just the radar number?

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -241,6 +242,17 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
     RetainPtr<CVMetalTextureCacheRef> m_coreVideoTextureCache;
 #endif
-};
+} SWIFT_SHARED_REFERENCE(retainDevice, releaseDevice);
```
I guess I don't understand what is the difference to passing ` WTF::retainThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr, WTF::releaseThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr` to the macro instead

mwyrzykowski replied:
then you don't need the custom inline functions in the header

nmahendru replied:
Interop doesn not support templated retain release functions. Hence the wrapper.

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -88,15 +88,18 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
+    void ensureBlitCommandEncoder();
+    void finalizeBlitCommandEncoder();
+
+    id<MTLBlitCommandEncoder> blitCommandEncoder() { RELEASE_ASSERT(m_blitCommandEncoder); return m_blitCommandEncoder; }
```
This will release assert when the Queue is invalid and called in the below Swift extension, but maybe `blitCommandEncoder` is never called on an invalid queue?

mwyrzykowski replied:
propose removing this `blitCommandEncoder` member function and following the pattern of CommandEncoder which has `id<MTLBlitCommandEncoder> ensureBlitCommandEncoder();`

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -88,15 +88,18 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
     static bool writeWillCompletelyClear(WGPUTextureDimension, uint32_t widthForMetal, uint32_t logicalSizeWidth, uint32_t heightForMetal, uint32_t logicalSizeHeight, uint32_t depthForMetal, uint32_t logicalSizeDepthOrArrayLayers);
     void endEncoding(id<MTLCommandEncoder>, id<MTLCommandBuffer>) const;
 
+    void ensureBlitCommandEncoder();
```
we could change the return type to `id<MTLBlitCommandEncoder>` similar to what CommandEncoder does so you don't need separate `ensure...` and accessor member functions

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -23,7 +23,32 @@
+        let device = self.device()
+        ensureBlitCommandEncoder()
+        let noCopy = data.size() >= largeBufferSize
+        precondition(device.device() != nil, "WeakPtr to Device is null")
```
This may fail, but I don't think we want it to abort. Is `precondition` the correct thing to use in that case?

rr-codes replied:
I think it should be changes to use an early return + assertionFailure (which is debug only)

mwyrzykowski replied:
If we assert, the tests validating this in debug would crash.

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -23,7 +23,32 @@
+                data.__dataUnsafe(), data.size(), MTLResourceOptions.storageModeShared)
+            : device.newBufferWithBytes(
+                data.__dataUnsafe(), data.size(), MTLResourceOptions.storageModeShared)
+        precondition(tempBuffer != nil, "TempBuffer is null. Should not be.")
```
similarly this allocation may fail, I've never looked at the documentation for Swift's precondition prior to today, but it sounds like it will abort

mwyrzykowski reviewed (changes requested):
Looks good overall, I have a few suggestions and I don't think we can use `precondition` in the Swift code unless I misunderstand how it works which is quite possible

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -124,13 +125,14 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
 
 } // namespace WebGPU
 
+// FIXME: rdar://137074301
```
same as the above comment on the FIXME

mwyrzykowski reviewed (approved):
Would propose adding bugzilla links with the FIXME, otherwise 👍

rr-codes commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -23,7 +23,39 @@
+}
+
+extension WebGPU.Queue {
+    public func writeBuffer(mtlBuffer: any MTLBuffer, bufferOffset: UInt64, data: SpanUInt8) {
```
can this use `some` instead of `any`?

nmahendru replied:
Good catch. Yes it can be `some`. Will do that in the next one.

nmahendru committed:
Swift Implementation for underlying Queue::writeBuffer
https://bugs.webkit.org/show_bug.cgi?id=280715
rdar://136672126

Reviewed by Mike Wyrzykowski.

Re-implement Queue::writeBuffer using a Swift implementation.
It is still using some underlying utilities from the C++ implementation
but that can only be moved over once Swift implements the whole class and has ownership
of underlying objects.

Testing:
Manual at desk builds with --webGpuSwift and running the Buffer and Queue tests.
Then Run writeBuffer.html test.

* Source/WTF/wtf/RefCounted.h:
(retainRefCounted): Deleted.
(releaseRefCounted): Deleted.
* Source/WTF/wtf/RetainReleaseSwift.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(retainThreadSafeRefCounted): Deleted.
(releaseThreadSafeRefCounted): Deleted.
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::retainThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
(WTF::releaseThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
(retainBuffer):
(releaseBuffer):
* Source/WebGPU/WebGPU/Device.h:
(retainDevice):
(releaseDevice):
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::blitCommandEncoder):
(retainQueue):
(releaseQueue):
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):
(writeBuffer(_:buffer:offset:data:)): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: https://commits.webkit.org/284638@main
